File: 91/go/91.go
// DP

// assume
// dp[i] is the possible ways of s[0~i]
// dp[i] =>
// 1. s[i-1 ~ i] is in ("1"~"26")  &&  s[i] is in ("1"~"26") => dp[i] = dp[i-1] + dp[i-2]
// 2. only s[i-1 ~ i] is in ("1"~"26") => dp[i] = dp[i-2]
// 3. only s[i] is in ("1"~"26") => dp[i] = dp[i-1]

import "strconv"

func numDecodings(s string) int {
    // check
    n := len(s)

    if s[0] == '0' {
        return 0
    }
    if n == 1 {
        return 1
    }

    // make dp
    dp := make([]int, n)
    // make checkM
    checkM := make(map[string]int)
    for i := 1; i <= 26; i++{
        checkM[strconv.Itoa(i)] = 1
    }

    // base cases
    dp[0] = 1
    _, find1 := checkM[s[1:2]]
    _, find2 := checkM[s[0:2]]
    if find1 && find2 {
        dp[1] = 2
    }else if find1 || find2{
        dp[1] = 1
    }

    if dp[1] == 0 {
        return 0
    }

    for i := 2; i < n; i++{
        _, find1 := checkM[s[i:i+1]]
        _, find2 := checkM[s[i-1:i+1]]
        if find1 && find2 {
            dp[i] = dp[i-1] + dp[i-2]
        }else if find1{
            dp[i] = dp[i-1]
        }else if find2{
            dp[i] = dp[i-2]
        }

        if dp[i] == 0 {
            return 0
        }
    }

    return dp[n-1]

}

File: 91/js/91.js
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
    // check
    if (s[0] === "0"){
        return 0;
    }
    let n = s.length;
    if (n === 1){
        return 1;
    }

    // assume preTwoBity start with empty ""
    // assume preOneBity start with s[0]
    let preOneBity = 1;
    let preTwoBity = 1;
    let res = 0;

    for (let i = 2; i <= n; i++){
        let num1 = s[i-1] - "0";
        let num2 = (s[i-2] - "0")*10 + num1;
        let find1 = (num1 > 0 && num1 < 10);
        let find2 = (num2 > 9 && num2 < 27);

        res = 0;

        if ( find1){
            res += preOneBity;
        }
        if (find2){
            res += preTwoBity;
        }

        if (res === 0){
            return 0;
        }

        preTwoBity = preOneBity;
        preOneBity = res;
    }
    return res;
};